File: python/src/devices.py
from _key import API_CLIENT_ID, SECRET
from _lib import generate_jwt_and_send_request

GRAPHQL_QUERY = """
query ($after: String) {
    devices(
        projectIds: [ "PLEASE ENTER YOUR PROJECT IDS HERE" ],
        after: $after
    )  {
        edges {
            node {
                id
                did
                macAddress
                os
                version
                lastSeen
                projectId
                createdAt
                deviceTags {
                    id
                        deviceId
                        deviceTagKey {
                            id
                            name
                        }
                        deviceTagValue
                }
            }
            cursor
        }
        pageInfo {
            hasNextPage
            endCursor
        }
    }
}
"""


def fetch_all_devices(api_client_id, secret):
    after = None
    all_devices = []

    while True:
        variables = {"after": after}

        response = generate_jwt_and_send_request(
            api_client_id, secret, GRAPHQL_QUERY, variables
        )

        devices_data = response.get("data", {}).get("devices", {})
        edges = devices_data.get("edges", [])
        page_info = devices_data.get("pageInfo", {})

        for edge in edges:
            all_devices.append(edge["node"])

        if not page_info.get("hasNextPage"):
            break

        after = page_info.get("endCursor")

    return all_devices


devices = fetch_all_devices(API_CLIENT_ID, SECRET)

print(devices)

File: nodejs/src/devices.js
const { generateJwtAndSendRequest } = require("./_lib");
const { apiClientId, secret } = require("./_key");

const GRAPHQL_QUERY = `
query ($after: String) {
    devices(
        projectIds: [ "PLEASE ENTER YOUR PROJECT IDS HERE" ],
        after: $after
    ) {
        edges {
            node {
                id
                did
                macAddress
                os
                version
                lastSeen
                projectId
                createdAt
                deviceTags {
                    id
                    deviceId
                    deviceTagKey {
                        id
                        name
                    }
                    deviceTagValue
                }
            }
            cursor
        }
        pageInfo {
            hasNextPage
            endCursor
        }
    }
}
`;

async function fetchAllDevices(apiClientId, secret) {
  let after;
  const allDevices = [];

  while (true) {
    const variables = { after };

    const response = await generateJwtAndSendRequest(
      apiClientId,
      secret,
      GRAPHQL_QUERY,
      variables
    );

    const devicesData = response?.data?.devices || {};
    const edges = devicesData.edges || [];
    const pageInfo = devicesData.pageInfo || {};

    for (const edge of edges) {
      allDevices.push(edge.node);
    }

    if (!pageInfo.hasNextPage) {
      break;
    }

    after = pageInfo.endCursor;
  }

  return allDevices;
}

(async () => {
  try {
    const devices = await fetchAllDevices(apiClientId, secret);
    console.log(JSON.stringify(devices));
  } catch (error) {
    console.error("Error fetching devices:", error);
  }
})();
